File: apps/server/src/services/container-service.ts
import { container, DependencyContainer } from 'tsyringe'

import {
  HealthCheckService,
  IHealthCheckService,
} from '../api/components/health-check'
import {
  IUserRepository,
  IUserService,
  UserRepository,
  UserService,
} from '../api/components/user'
import { config } from '../config/config'
import { DBService, IDBService } from '../services/db-service'
import { IPaginationService, PaginationService } from './pagination-service'

class ContainerService {
  private readonly _container: DependencyContainer

  constructor() {
    this._container = container
    this.registerDatabaseContainer()
    this.registerRepositoryContainer()
    this.registerServiceContainer()
  }

  public get container(): DependencyContainer {
    return this._container
  }

  private registerDatabaseContainer() {
    const dbService = new DBService()
    dbService.connectDatabase(config.getDatabaseURL())
    this._container.register<IDBService>('DBService', {
      useValue: dbService,
    })
  }

  private registerRepositoryContainer() {
    this._container.register<IUserRepository>('UserRepository', {
      useValue: new UserRepository(
        this._container.resolve<DBService>('DBService'),
      ),
    })
  }

  private registerServiceContainer() {
    this._container.register<IHealthCheckService>('HealthCheckService', {
      useValue: new HealthCheckService(
        this._container.resolve<DBService>('DBService'),
      ),
    })
    this._container.register<IUserService>('UserService', {
      useValue: new UserService(
        this._container.resolve<UserRepository>('UserRepository'),
      ),
    })
    this._container.register<IPaginationService>('PaginationService', {
      useValue: new PaginationService(),
    })
  }
}

export { ContainerService }

File: apps/server2/src/services/container_service.py
from dependency_injector import containers, providers

from src.api.components.health.health_check_service import HealthCheckService
from src.services.db_service import DatabaseService

# async def session_factory(conn_string: str):
#     database_session_manager = DatabaseSessionManager(conn_string=conn_string)
#     async with database_session_manager.session() as session:
#         yield session


class Container(containers.DeclarativeContainer):
    db_service: providers.Factory
    health_check_service: providers.Factory

    def __init__(self):
        self.db_service = None
        self.health_check_service = None


class ContainerService:
    __container: Container

    def __init__(self):
        self.__container = Container()
        self.register_database_container()
        # self.register_repository_container()
        self.register_service_container()

    @property
    def container(self) -> containers.DeclarativeContainer:
        return self.__container

    def register_database_container(self):
        self.__container.db_service = providers.Factory(DatabaseService)

    def register_service_container(self):
        self.__container.health_check_service = providers.Factory(
            HealthCheckService, db_service=self.__container.db_service
        )
